cypress/e2e/permit2.test.ts

```diff
@@ -0,0 +1,206 @@
+    cy.visit(`/swap/?inputCurrency=${INPUT_TOKEN.address}&outputCurrency=${OUTPUT_TOKEN.address}`, {
+      ethereum: 'hardhat',
+    })
+    cy.get('#swap-currency-input .token-amount-input').clear().type(TEST_BALANCE_INCREMENT.toString())
+  })
+
+  /** Asserts permit2 has a max approval for spend of the input token on-chain. */
+  const expectTokenAllowanceForPermit2ToBeMax = () => {
+    // check token approval
+    return cy
+      .hardhat()
+      .then(({ approval, wallet }) => approval.getTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN }))
+      .should('deep.equal', MaxUint256)
+  }
+
+  /** Asserts the universal router has a max permit2 approval for spend of the input token on-chain. */
+  const expectPermit2AllowanceForUniversalRouterToBeMax = (approvalTime: number) => {
+    return cy
+      .hardhat()
+      .then((hardhat) => hardhat.approval.getPermit2Allowance({ owner: hardhat.wallet, token: INPUT_TOKEN }))
+      .then((allowance) => {
+        cy.then(() => MaxUint160.eq(allowance.amount)).should('eq', true)
+        const expected = Math.floor((approvalTime + 2_592_000_000) / 1000)
+
+        // Asserts that the on-chain expiration is in 30 days, within a tolerance of 20 seconds.
+        cy.then(() => Math.abs(allowance.expiration - expected)).should('be.lessThan', 20)
+      })
+  }
+
+  it('does not prompt approval when user has already approved token and permit2', () => {
+    cy.hardhat()
+      .then(({ approval, wallet }) => {
+        approval.setTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN })
+        approval.setPermit2Allowance({ owner: wallet, token: INPUT_TOKEN })
+      })
+      .then(swaps)
+  })
+
+  it('can swap after completing full permit2 approval process', () => {
+    cy.get(APPROVE_BUTTON).click()
+    const approvalTime = Date.now()
+    cy.get(APPROVE_BUTTON).should('have.text', 'Approval pending')
+
+    // There should be a successful Approved notification.
+    cy.contains('Approved').should('exist')
+
+    swaps()
+
+    expectTokenAllowanceForPermit2ToBeMax()
+    expectPermit2AllowanceForUniversalRouterToBeMax(approvalTime)
+  })
+
+  it('can swap after handling user rejection of approvals and signatures', () => {
+    const USER_REJECTION = { code: 4001 }
+    cy.hardhat().then((hardhat) => {
+      const tokenApprovalStub = cy.stub(hardhat.wallet, 'sendTransaction')
+      tokenApprovalStub.rejects(USER_REJECTION) // reject token approval
+
+      // Clicking the approve button should trigger a token approval that will be rejected by the user (tokenApprovalStub).
+      cy.get(APPROVE_BUTTON).click()
+
+      // The swap component should prompt approval again.
+      cy.get(APPROVE_BUTTON)
+        .should('have.text', `Approve use of ${INPUT_TOKEN.symbol}`)
+        .then(() => {
+          const permitApprovalStub = cy.stub(hardhat.provider, 'send')
+          permitApprovalStub.withArgs('eth_signTypedData_v4').rejects(USER_REJECTION) // reject permit approval
```

Comment:
```suggestion
          permitApprovalStub.withArgs('eth_signTypedData_v4').rejects(USER_REJECTION) // reject permit approval
          permitApprovalStub.callThrough() // <- this tells it to just call the orig. impl. if the args don't match, otherwise it will stub *everything*
```

Reply:
thank you! much better than the answer the bot replied to my PR with above